### PR TITLE
refactor: Extract `Store::queryPathInfoFromClientCache`

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

This "new" method is useful for quickly figuring out whether a store is known to have a path. This will be useful to me as a C++ library consumer. This does not add new call sites - it is just a refactor that exposes a "new" method.

> [!TIP]
> This refactor is split into three small steps with separate commits, if that helps you see the equivalence.
> Disable whitespace in the diff.

# Context

An alternative would be for users to invoke the narinfo cache db directly,
so why should we make this change?

 - This way, the narinfo cache db remains an implementation detail.
 - Callers get to use the in-memory cache as well.
 - It is easier to use.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Review by Ericson2314 — APPROVED** [review]

File: src/libstore/store-api.cc [hunk 1]
```diff
@@ -686,38 +686,56 @@ static bool goodStorePath(const StorePath & expected, const StorePath & actual)
...
+    try {
+        auto r = queryPathInfoFromClientCache(storePath);
+        if (r.has_value()) {
+            std::shared_ptr<const ValidPathInfo> & info = *r;
+            if (info)
+                return callback(ref(info));
+            else
+                throw InvalidPath("path '%s' is not valid", printStorePath(storePath));
+        }
     } catch (...) { return callback.rethrow(); }
```
Comment (thufschmitt) [inline src/libstore/store-api.cc:739]: (nit) do we still need the try-catch? Can't we have something like below instead?

```suggestion
{
    auto hashPart = std::string(storePath.hashPart());

    auto r = queryPathInfoFromClientCache(storePath);
    if (r.has_value()) {
        std::shared_ptr<const ValidPathInfo> & info = *r;
        if (info)
            return callback(ref(info));
        else
            return callback.rethrow(std::make_exception_ptr<InvalidPath>("path '%s' is not valid", printStorePath(storePath)));
    }
```

**Review by thufschmitt — APPROVED** [review]
One small potential simplification, but looks good otherwise

File: src/libstore/store-api.cc (hunk 1 above)
Comment (roberth) [reply src/libstore/store-api.cc:739]: It does also catch other exceptions, which is important to the contract.
Although I've removed the unchecked `noexcept` modifier (I don't think exceptions should be "upgraded" to crashes), it's still expected for Callback based functions not to throw.

**Comment by roberth** [comment]
I've fixed up a small mistake in the middle commit just now, but the final tree is identical.

**Pull request merged**